astropy/wcs/src/wcslib_wrap.c

```diff
@@ -157,6 +159,135 @@ convert_rejections_to_warnings() {
   return status;
 }
 
+
+/***************************************************************************
+ * wtbarr-related global variables and functions                           *
+ ***************************************************************************/
+
+static PyObject *get_wtbarr_data = NULL;
+
+
+void _set_wtbarr_callback(PyObject* callback) {
+  Py_XINCREF(callback);         /* Add a reference to new callback */
+  Py_XDECREF(get_wtbarr_data);  /* Dispose of previous callback */
+  get_wtbarr_data = callback;   /* Remember new callback */
+}
+
+
+int _update_wtbarr_from_hdulist(PyObject *hdulist, struct wtbarr *wtb) {
+  PyArrayObject *arrayp=NULL;
+  PyObject *result=NULL, *arglist = NULL;
+  int i, naxis, nelem, naxes[NPY_MAXDIMS];
+  npy_intp *npy_naxes;
+  npy_double *appayp_data;
+
+  if (hdulist == NULL || hdulist == Py_None) {
+    PyErr_SetString(PyExc_ValueError,
+                    "HDUList is required to retrieve -TAB coordinates "
+                    "and/or indices.");
+    return 0;
+  }
+
+  if (wtb->ndim < 1) {
+    PyErr_SetString(PyExc_ValueError, "Number of dimensions should be positive.");
+    return 0;
+  }
+
+  result = PyObject_CallFunction(get_wtbarr_data, "(OsiiCsii)", hdulist,
+      wtb->extnam, wtb->extver, wtb->extlev, wtb->kind, wtb->ttype, wtb->row,
+      wtb->ndim);
+
+  if (result == NULL) return 0;
+
+  arrayp = (PyArrayObject *)PyArray_FromAny(result,
+      PyArray_DescrFromType(NPY_DOUBLE), 0, 0, NPY_ARRAY_CARRAY, NULL);
+
+  if (arrayp == NULL) {
+    PyErr_SetString(PyExc_TypeError, "Unable to convert wtbarr callback "
+                    "result to a numpy.ndarray.");
+    Py_DECREF(result);
+    return 0;
+  }
+
+  Py_DECREF(result);
+
+  if (!PyArray_Check(arrayp)) {
+    PyErr_SetString(PyExc_TypeError,
+                    "wtbarr callback must return a numpy.ndarray type "
+                    "coordinate or index array.");
+    Py_DECREF(arrayp);
+    return 0;
+  }
+
+  naxis = PyArray_NDIM(arrayp);
+
+  if (naxis == 0) {
+    PyErr_SetString(PyExc_ValueError, "-TAB coordinate or index arrays "
+                    "cannot be 0-dimensional.");
+    Py_DECREF(arrayp);
+    return 0;
+  }
+
+  npy_naxes = PyArray_DIMS(arrayp);
+  for (i = 0; i < naxis; i++) {
+    naxes[i] = (int) npy_naxes[i];
+  }
+
+  if (naxis != wtb->ndim) {
+    if (wtb->kind == 'c' && wtb->ndim == 2 && naxis == 1) {
+      /* Allow TDIMn to be omitted for degenerate coordinate arrays. */
+      naxis = 2;
+      naxes[1] = 1;
+    } else {
+      PyErr_Format(PyExc_ValueError,
+          "An array with an unexpected number of axes was "
+          "received from the callback. Expected %d but got %d.",
+          wtb->ndim, (int) naxis);
+      Py_DECREF(arrayp);
+      free(naxes);
```

Comment:
Since `naxes` is stack-allocated this will be undefined behavior (see also later).